lib/ide/parser/src/api.rs

```diff
@@ -8,9 +9,11 @@ pub type Ast = ast::Ast;
 // == Parser ==
 // ============
 
+pub type IDMap = Vec<((isize, isize), ID)>;
```

Comment:
I'd wrap this in a strong type, current usage of IDMap should not rely on it being `Vec`.
(actually, no one really uses it in other way then serializing and passing to JS — when we will use it, then we will actually know what underlying container works best)

Reply:
I don't see how wrapping it removes the usage of `Vec`. 

I assume the wrapper would be struct, so we would just replace `vec![..]` with more verbose `IDMap { vec![] }`.

Reply:
Huh. To correctly hide the vec, one would need top provide construction methods, like `new` or  FromIterator trait.

But I've been thinking about this.

Actually what I don't like is weak typing (by weak typing I mean exposing raw integers and, most relevantly, tuples, instead of named types/fields).  Basically the shape of this IDMap is incidental, created to match the serialziation format for Scala interop, not to express data semantics.

The "proper" ID Map should not really expose the underlying storage, rather hide it behind some API. Also, there should be no tuples like (isize, isize) but we should solve it like we did in Scala — have proper struct for Position/Offset/Span. Having strong types also allows us providing methods that'll make usage much nicer.

But now I think we actually need two ID Map types:
1. the "proper", strongly typed one (like the one I'm describing)
2. the JS serialziation compatible one (like the one you have here).

The first one should be the one exposed in parser API. 
The second one should be created only incidentally, for serialization, when passing idmap to scala. It should not be part of user-targeted interface of this crate.

As part of this PR, I'm happy with having 1. as total stub (e.g. being only able to use default() to construct empty map). Its development will be likely driven by other tasks, like generating idmap from ast or persisting it in module file. And, most importantly, text API.

2. in its current shape is actually fine but its usage should be limited to rust/scala interop.